File: doric-android/doric/src/main/java/pub/doric/loader/DoricJSLoaderManager.java
/*
 * Copyright [2019] [Doric.Pub]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package pub.doric.loader;


import android.net.Uri;
import android.text.TextUtils;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

import pub.doric.DoricSingleton;
import pub.doric.async.AsyncResult;

/**
 * @Description: pub.doric
 * @Author: pengfei.zhou
 * @CreateDate: 2019-11-23
 */
public class DoricJSLoaderManager {

    private final Set<IDoricJSLoader> jsLoaders = new HashSet<>();

    public DoricJSLoaderManager() {
        addJSLoader(new DoricAssetJSLoader());
        addJSLoader(new DoricHttpJSLoader());
    }

    public void addJSLoader(IDoricJSLoader jsLoader) {
        jsLoaders.add(jsLoader);
    }

    private Collection<IDoricJSLoader> getJSLoaders() {
        return jsLoaders;
    }

    public AsyncResult<String> loadJSBundle(String source) {
        if (!TextUtils.isEmpty(source)) {
            if (source.startsWith("_internal_://")) {
                Uri uri = Uri.parse(source);
                String srcContextId = uri.getQueryParameter("context");
                String className = uri.getQueryParameter("class");
                source = String.format("Entry('%s','%s')", srcContextId, className);
                return new AsyncResult<>(source);
            }
            Collection<IDoricJSLoader> jsLoaders = getJSLoaders();
            for (IDoricJSLoader jsLoader : jsLoaders) {
                if (jsLoader.filter(source)) {
                    return jsLoader.request(source);
                }
            }
        }
        AsyncResult<String> ret = new AsyncResult<>();
        ret.setError(new RuntimeException("Cannot find JS Loader for " + source));
        return ret;
    }

    @Deprecated
    public static DoricJSLoaderManager getInstance() {
        return DoricSingleton.getInstance().getJSLoaderManager();
    }
}

File: doric-flutter/doric/lib/doric/loader/doric_JSLoaderManager.dart
import 'dart:collection';

import 'package:doric/doric/loader/doric_asset_jsloader.dart';
import 'package:doric/doric/loader/idoric_JSLoader.dart';

class DoricJSLoaderManager {
  Set<IDoricJSLoader> _jsloader = new HashSet();

  DoricJSLoaderManager._(){
    addJSLoader(AssetJSLoader());
  }

  static final _instance = DoricJSLoaderManager._();

  factory DoricJSLoaderManager.getInstance() => _instance;
  void addJSLoader(IDoricJSLoader jsLoader) {
    _jsloader.add(jsLoader);
  }

  Set<IDoricJSLoader> getJSLoaders() {
    return _jsloader;
  }

  Future<String> loadJSBundle(String source) {
    if (source != null && source.isNotEmpty) {
      for (IDoricJSLoader loader in getJSLoaders()) {
        if (loader.filter(source)) {
          return loader.request(source);
        }
      }
    }
    return null;
  }
}
